File: client/src/app/models/map-marker.ts
import {AbstractModel} from './abstract-model';
import {MapMarkerType} from '../enums/map-marker-type';
import {Coordinates} from '../interfaces/coordinates.interface';
import {Crag} from './crag';
import {Sector} from './sector';
import {Area} from './area';
import {TopoImage} from './topo-image';

export type MapMarkerProperties = {
  name: string;
  description: string;
  type: MapMarkerType;
  crag: Partial<Crag>;
  sector: Partial<Sector>;
  area: Partial<Area>;
  topoImage: Partial<TopoImage>;
};

export class MapMarker extends AbstractModel {
  name: string;
  description: string;
  coordinates: Coordinates;
  customIcon: string;
  color: string;
  type: MapMarkerType;

  public static deserialize(payload: any): MapMarker {
    const mapMarker = new MapMarker();
    AbstractModel.deserializeAbstractAttributes(mapMarker, payload);
    mapMarker.name = payload.name;
    mapMarker.description = payload.description;
    mapMarker.coordinates = { lat: payload.lat, lng: payload.lng };
    mapMarker.customIcon = payload.customIcon;
    mapMarker.color = payload.color;
    mapMarker.type = payload.type;
    return mapMarker;
  }

  public static serialize(mapMarker: MapMarker): any {
    return {
      name: mapMarker.name,
      description: mapMarker.description,
      lat: mapMarker.coordinates.lat,
      lng: mapMarker.coordinates.lng,
      customIcon: mapMarker.customIcon,
      color: mapMarker.color,
      type: mapMarker.type,
    };
  }
}

File: server/src/models/map_marker.py
from extensions import db
from models.base_entity import BaseEntity

from models.enums.map_marker_type_enum import MapMarkerType
from sqlalchemy.dialects.postgresql import UUID



class MapMarker(BaseEntity):
    """
    Model of a map marker. Can mark anything from crag to parking spot.
    """
    __tablename__ = 'map_markers'

    lat = db.Column(db.Float, nullable=True)
    lng = db.Column(db.Float, nullable=True)
    type = db.Column(db.Enum(MapMarkerType), nullable=False)
    name = db.Column(db.String(120), nullable=True)
    description = db.Column(db.Text, nullable=True)
    # todo cascade
    crag_id = db.Column(UUID(as_uuid=True), db.ForeignKey('crags.id'), nullable=True)
    sector_id = db.Column(UUID(as_uuid=True), db.ForeignKey('sectors.id'), nullable=True)
    area_id = db.Column(UUID(as_uuid=True), db.ForeignKey('areas.id'), nullable=True)
    topo_image_id = db.Column(UUID(as_uuid=True), db.ForeignKey('topo_images.id'), nullable=True)

    crag = db.relationship('Crag', back_populates='map_markers')
    sector = db.relationship('Sector', back_populates='map_markers')
    area = db.relationship('Area', back_populates='map_markers')
    topo_image = db.relationship('TopoImage', back_populates='map_markers')

